### PR TITLE
refactor: rewrite root dir by using vim.fs module

### PR DESCRIPTION
**Pull request by glepnir**
a break change . the `vim.fs` module only works for nvim 0.8+ . so idk what time is can be merge.

relate issue: #2449  #2400 #1908

### CODE REVIEW

**Comment by glepnir** [comment]
I want use `vim.fs` module to rewrite root dir section. but the `vim.fs` module only support 0.8+ .

@mikehaertl  can you test this ?

**Comment by mikehaertl** [comment]
@glepnir Actually I don't have a use case for #1908 anymore. It was related to setting up the LSP for volar but since I migrated to mason I no longer have to deal with it.

Anyway, as described in the issue testing it is easy:

```
mkdir -p /home/mike/test/a/node_modules
mkdir /home/mike/test/node_modules
```
Then in neovim:

```
:lua print( require'lspconfig'.util.find_node_modules_ancestor('/home/mike/test/a') )
```

Output should be `/home/mike/test` not `/home/mike/test/a`.

**Comment by glepnir** [comment]
i want rewrite root search in this pr. but ikd  what time this pr can be merge lol

**Comment by mikehaertl** [comment]
I can't help much. But the method is called from many places so the change might break things.

**Comment by glepnir** [comment]
I had rewrite it and some change is breakchange. now should works .parent will be root in some situation.

**Comment by mikehaertl** [comment]
Note that `search_ancestors()` is also used in some code snippets in the docs, e.g.:

https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#volar

File: lua/lspconfig/util.lua [hunk 1]
```diff
@@ -453,38 +446,45 @@ function M.root_pattern(...)
...
 end
+
 function M.find_node_modules_ancestor(startpath)
-  return M.search_ancestors(startpath, function(path)
-    if M.path.is_dir(M.path.join(path, 'node_modules')) then
-      return path
-    end
-  end)
+  return function()
+    return searcher(startpath, { '.hg' }, true)
```
Comment (mikehaertl) [inline lua/lspconfig/util.lua:480]:
```suggestion
    return searcher(startpath, { 'node_modules' }, true)
```

File: lua/lspconfig/util.lua [hunk 2]
```diff
@@ -453,38 +446,45 @@ function M.root_pattern(...)
...
-    if M.path.is_dir(M.path.join(path, '.git')) or M.path.is_file(M.path.join(path, '.git')) then
-      return path
-    end
-  end)
+  return function()
+    return searcher(startpath, { '.git' })
+  end
 end
+
 function M.find_mercurial_ancestor(startpath)
```
Comment (mikehaertl) [inline lua/lspconfig/util.lua:472]: Maybe also add a `tail` param here and to the other helpers that defaults to `false`? This would not break bc.

Comment (glepnir) [reply lua/lspconfig/util.lua:472]: yep . just for test i set it to true. just search on github. found lots of function in utils are used in people config ... can't do break change ..

Comment (glepnir) [reply lua/lspconfig/util.lua:472]: Needs more consideration and design here. take compatibility issues into consideration.  🧐

**Pull request closed without merging**